README.md
# Contemporary Controls API Interface for Universal Device Polyglot NodeServers

## bascontrolns

![BASpi-Edge Controller](https://github.com/sjpbailey/bascontrol_ns/blob/9cd9ebeed425590cc86aa60a54a488f71171b614/archive%2Fbasedge.jpg)

## Based on the Contemporary Controls BASpi, BASpi-Edge DIY and BAScontrol Series Controllers

## BASpi-Edge & BASpi 6U6R/6U4R2AO, BAScontrol DIY Series BacNet Control Devices by Contemporary Controls information

* Main
[Contemporary Controls BASpi DIY](https://www.ccontrols.com/basautomation/baspi.htm)
* BASpi 6U6R Controller
[Contemporary Controls BASpi 6U6R](https://www.ccontrols.com/pdf/ds/BASPI-datasheet.pdf)
* BASpi 6U6R Installation
[Contemporary Controls BASpi 6U6R Install](https://www.ccontrols.com/pdf/BASpi-hardware-install-guide.pdf)
* BASpi 6U4R2A Controller
[Contemporary Controls BASpi 6U4R2A](https://www.ccontrols.com/pdf/ds/BASPI-AO2-datasheet.pdf)
* BASpi 6U4R2A Installation
[Contemporary Controls BASpi 6U4R2A Install](https://www.ccontrols.com/pdf/TD180600.pdf)
* BASpi Controller Configuration
[Contemporary Controls BASpi Configuration Quick Start](https://www.ccontrols.com/pdf/is/BASPI-QSGuide.pdf)

### Installation

* Installation pip install: [bascontrolns](https://pypi.org/project/bascontrolns/)

#### Requirments

* Python 3.7

* requests

* sjb, gtb


CHANGELOG.md
# Change Log for bascontrolns Interface for Contemporary Controls

## Version 0.0.3

- Version 0.0.3: This includes BASpi-All, BASpi-Edge 6UI/6R and the BAScontrol series
- Future release will include BAS-Edge 6UI/4R/2AO
- working on new bascontrolns adding BASpi-Edge 6UI/4R/2AO

### Version 0.0.4 June 2nd 2021

- Version 0.0.4 Added abillity to write to Sedona Wire Sheet nodes or points
- Modified Virtual Values to control Nodes in Sedona Wire Sheet


bascontrolns/bascontrolns.py
# For HTML POST requests.
import requests
# For Python 3.x enumerated types.
from enum import Enum
# For XML parsing.
import xml.etree.ElementTree as et
# For time functions.
import time


###############################################################################
#
class Platform(Enum):

    """BAScontrol platform (hardware) types"""

    BASC_NONE = 0
    BASC_20  = 0
    BASC_PI  = 1
    BASC_AO  = 2
    BASC_PO  = 3
    BASC_ED  = 4


###############################################################################
#
def getUrl(sIpAddress):

    """Returns the full cgi URL of the target"""

    return 'http://' + sIpAddress + '/cgi-bin/xml-cgi'


###############################################################################
#
def getModel(url , secTimeout):

    """Uses RDOM to get the model of the target (BAScontrol22, etc.)"""

    model = 'Error'
    req = '<rdom fcn="get" doc="sts" path="model"/>'
    try:
        rsp = requests.post(url , data = req , timeout = secTimeout)
        if (rsp.status_code == 200):
            root = et.fromstring(rsp.text)
            if root.attrib['rsp'] == 'ack':
                model = root.text
    except OSError:
        # Timeout.
        pass
    return model


###############################################################################
#
def getPlatform(sModel):

    """Returns the Platform type given the model (BASconrol22, etc.)"""

    eRetval = Platform.BASC_NONE
    if sModel.find('20') != -1:
        eRetval = Platform.BASC_20
    elif sModel.find('6R') != -1:
        eRetval = Platform.BASC_PI
    elif sModel.find('PI') != -1:
        eRetval = Platform.BASC_PO
    elif sModel.find('2AO') != -1:
        eRetval = Platform.BASC_AO 
    elif sModel.find('6') != -1:
        eRetval = Platform.BASC_ED       
    return eRetval


###############################################################################
#
# BAScontrol device class.
# Uses RDOM to get/set the present value for all objects on the target that
# are part of the configuration file; UI, BI, AO, BO, and VT's.
#
class Device:

    """BAScontrol device class: Initialize with IP address"""

    def __init__(self , sIpAddress , bInit=True):

        # Assume failure.
        self.online = False
        self.ePlatform = Platform.BASC_NONE
        self.sModel = 'Error'
        # Assign complete IP address.
        self.sFullUrl = getUrl(sIpAddress)

        # Default timeout and retry interval.
        self.timeout = 2
        self.retryInterval = 15
        self.nextRetry = 0

        # Assign default values.
        self.vtQty = 24
        self.uiQty = 0
        self.biQty = 0
        self.aoQty = 0
        self.boQty = 0
        #
        self.uiBase = 0
        self.biBase = 0
        self.aoBase = 0
        self.boBase = 0
        self.vtBase = 0

        # Return if we're not initializing right away.
        if not bInit:
            return
        # Set up the device.
        self.initialize()

    ###########################################################################
    #
    def initialize(self):

        """Checks online status and does setup()"""

        # Is target online?
        if self.isOnline():
            # Setup our data.
            self.setup()
            return True

        # Not online.
        return False

    ###########################################################################
    #
    def isOnline(self):

        """Returns True if the target is online, False otherwise"""

        if self.online:
            return True

        self.sModel = getModel(self.sFullUrl , self.timeout)
        if self.sModel == 'Error':
            self.online = False
            self.ePlatform = Platform.BASC_NONE
            self.retryInterval = int(time.time())+self.retryInterval
        else:
            self.online = True
            self.ePlatform = getPlatform(self.sModel)
            self.retryInterval = 0
        return self.online

    ###########################################################################
    #
    def setup(self):

        """Sets device object quantities per the model"""

        # Get target platform.
        self.ePlatform = getPlatform(self.sModel)
        # Assign the number of IO.
        if self.ePlatform == Platform.BASC_20:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
        elif self.ePlatform == Platform.BASC_PI:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24
        elif self.ePlatform == Platform.BASC_AO:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 2
            self.boQty = 4
        elif self.ePlatform == Platform.BASC_PO:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24
        elif self.ePlatform == Platform.BASC_ED:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24        
            # Assign 0-based index values.
        self.uiBase = 0
        self.biBase = self.uiQty
        self.aoBase = self.biBase + self.biQty
        self.boBase = self.aoBase + self.aoQty
        self.vtBase = self.boBase + self.boQty

    ###########################################################################
    #
    def readObject(self , iIndex):

        """Returns response text if success, None if failure"""

        # If not online.
        if not self.online:
            # If retry period has expired.
            if int(time.time()) >= self.nextRetry:
                # If we're still offline.
                if not self.isOnline():
                    # New retry time has been set.
                    return None
            # Else need to wait.
            else:
                return None

        # Assume failure.
        retval = None
        # XML read request string.
        req = '<rdom fcn="get" doc="rtd" path="obj[' + str(iIndex) + ']/val_scl"/>'
        # Try to execute the POST.
        try:
            rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
            # If good HTML response.
            if (rsp.status_code == 200):
                root = et.fromstring(rsp.text)
                if root.attrib['rsp'] == 'ack':
                    retval = root.text
            else:
                # Something there, but not BAScontrol.
                self.online = False
        except OSError:
            # Timeout.
            self.online = False
        # If we were taken offline.
        if not self.online:
            # Set for retry.
            self.nextRetry = int(time.time())+self.retryInterval
            # Raise an exception.
            ###raise IOError
            # NOT REACHED
        # And return.
        return retval 

    ###########################################################################
    #
    # This does two POST operations:
    #   1. Checks ws_control to see if it's under wire sheet control.
    #   2. Executes the write if not under wire sheet control.
    #
    def writeObject(self , iIndex , value):

        """Writes the present value of an object using 0-based index"""

        # If not online.
        if not self.online:
            # If retry period has expired.
            if int(time.time()) >= self.nextRetry:
                # If we're still offline.
                if not self.isOnline():
                    # New retry time has been set.
                    return -1
            # Else need to wait.
            else:
                return -1

        # Assume failure.
        retval = -1
        # XML wire sheet control request string.
        req = '<rdom fcn="get" doc="rtd" path="obj[' + str(iIndex) + ']/ws_control"/>'
        # Try to execute first POST request.
        try:
            # Execute POST and get response.
            rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
            # If good HTML response.
            if (rsp.status_code == 200):
                # Parse XML.
                root = et.fromstring(rsp.text)
                # If request was ACK'd.
                if root.attrib['rsp'] == 'ack':
                    # If point is under wire sheet control.
                    if root.text == '1':
                        # We can't write.
                        return -1
                # Else we were NAK'd; SERIOUS PROBLEM
                else:
                    return -1
                #
                # If we get here, we can proceed with the write operation.
                #
                # XML write request string.
                req = '<rdom fcn="set" doc="rtd" path="obj[' + str(iIndex) + ']/val_scl">' + str(value) + '</rdom>'
                # Try to execute POST request.
                try:
                    # Execute POST and get response.
                    rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
                    # If good HTML response.
                    if (rsp.status_code == 200):
                        # Parse XML.
                        root = et.fromstring(rsp.text)
                        #######################################################
                        # This is the only place we can return success.
                        #######################################################
                        if root.attrib['rsp'] == 'ack':
                            retval = 0
                    else:
                        # Something there, but not BAScontrol.
                        self.online = False
                except OSError:
                    # Timeout.
                    self.online = False
            else:
                # Something there, but not BAScontrol.
                self.online = False
        except OSError:
            # Timeout.
            self.online = False
        # If we were taken offline.
        if not self.online:
            # Set for retry.
            self.nextRetry = int(time.time())+self.retryInterval
            # Raise an exception.
            ###raise IOError
            # NOT REACHED
        # And return.
        return retval

    ###########################################################################
    #
    def universalInput(self , iIndex):

        """Reads a universal input using 1-based index (UI1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.uiQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1)
        # Execute read.
        value = self.readObject(index)
        # And return.
        return None if value == None else float(value)
        

    ###########################################################################
    #
    def binaryInput(self , iIndex):

        """Reads a binary input using 1-based index (BI1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.biQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.biBase)
        # Execute read.
        value = self.readObject(index)
        # And return.
        return None if value == None else int(value)

    ###########################################################################
    #
    def analogOutput(self , iIndex , fValue = None):

        """Reads or writes an analog output using 1-based index (AO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.aoQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.aoBase)
        # If reading.
        if fValue == None:
            # Execute read.
            value = self.readObject(index)
            # And return.
            return None if value == None else float(value)
        # Else writing.
        return self.writeObject(index , fValue)

    ###########################################################################
    #
    def binaryOutput(self, iIndex, iValue=None):

        """Reads or writes an binary output using 1-based index (BO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.boQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.boBase)
        # If reading.
        if iValue == None:
            # Execute read.
            value = self.readObject(index)
            # Return if failure.
            if value == None:
                return None
            # Return with integer value.
            return 0 if int(value) == 0 else 1
        # Else writing.
        return self.writeObject(index, iValue)

    ###########################################################################
    #
    def virtualValue(self , iIndex , bIsBinary , aValue = None):

        """Reads or writes a virtual point using 1-based index (BO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.vtQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.vtBase)
        # If reading.
        if aValue == None:
            # Execute read.
            value = self.readObject(index)
            # Return if failure.
            if value == None:
                return None
            # Return with value.
            return int(value) if bIsBinary else float(value)
        # Else writing.
        return self.writeObject(index , aValue)


bascontrolns/bascontrolns_new-edge-added.py
# For HTML POST requests.
import requests
# For Python 3.x enumerated types.
from enum import Enum
# For XML parsing.
import xml.etree.ElementTree as et
# For time functions.
import time


###############################################################################
#
class Platform(Enum):

    """BAScontrol platform (hardware) types"""
    # Platform ID             Platform Response   
    BASC_NONE = 0
    BASC_20  = 0 ## BASControl Series I20 I20C I22 I22C
    BASC_PI  = 1 # '<rdom rsp="ack">BASpi 6U/6R</rdom>'
    BASC_AO  = 2 # '<rdom rsp="ack">BASpi 6U/4R/2AO</rdom>'
    BASC_PO  = 3
    BASC_ED  = 4 # '<rdom rsp="ack">BASpi-Edge 6/6</rdom>'
    BASC_EO  = 5 # from BASemulator BASpi-Edge 6/4/2


###############################################################################
#
def getUrl(sIpAddress):

    """Returns the full cgi URL of the target"""

    return 'http://' + sIpAddress + '/cgi-bin/xml-cgi'


###############################################################################
#
def getModel(url , secTimeout):

    """Uses RDOM to get the model of the target (BAScontrol22, etc.)"""

    model = 'Error'
    req = '<rdom fcn="get" doc="sts" path="model"/>'
    try:
        rsp = requests.post(url , data = req , timeout = secTimeout)
        if (rsp.status_code == 200):
            root = et.fromstring(rsp.text)
            if root.attrib['rsp'] == 'ack':
                model = root.text
    except OSError:
        # Timeout.
        pass
    return model


###############################################################################
#
def getPlatform(sModel):

    """Returns the Platform type given the model (BASconrol22, etc.)"""

    eRetval = Platform.BASC_NONE
    if sModel.find('20') != -1:      ## BASControl Series I20 I20C I22 I22C
        eRetval = Platform.BASC_20
    elif sModel.find('6R') != -1:    ## '<rdom rsp="ack">BASpi 6U/6R</rdom>'
        eRetval = Platform.BASC_PI
    elif sModel.find('PI') != -1:    ### '<rdom rsp="ack">BASpi 6U/6R</rdom>' original PI
        eRetval = Platform.BASC_PO
    elif sModel.find('2AO') != -1:   ## '<rdom rsp="ack">BASpi 6U/4R/2AO</rdom>'
        eRetval = Platform.BASC_AO 
    elif sModel.find('6') != -1:
        eRetval = Platform.BASC_ED   ## '<rdom rsp="ack">BASpi-Edge 6/6</rdom>'
    elif sModel.find('2') != -1:
        eRetval = Platform.BASC_EO   ## BASpi-Edge 6/4/2           
    return eRetval                   


###############################################################################
#
# BAScontrol device class.
# Uses RDOM to get/set the present value for all objects on the target that
# are part of the configuration file; UI, BI, AO, BO, and VT's.
#
class Device:

    """BAScontrol device class: Initialize with IP address"""

    def __init__(self , sIpAddress , bInit=True):

        # Assume failure.
        self.online = False
        self.ePlatform = Platform.BASC_NONE
        self.sModel = 'Error'
        # Assign complete IP address.
        self.sFullUrl = getUrl(sIpAddress)

        # Default timeout and retry interval.
        self.timeout = 2
        self.retryInterval = 15
        self.nextRetry = 0

        # Assign default values.
        self.vtQty = 24
        self.uiQty = 0
        self.biQty = 0
        self.aoQty = 0
        self.boQty = 0
        #
        self.uiBase = 0
        self.biBase = 0
        self.aoBase = 0
        self.boBase = 0
        self.vtBase = 0

        # Return if we're not initializing right away.
        if not bInit:
            return
        # Set up the device.
        self.initialize()

    ###########################################################################
    #
    def initialize(self):

        """Checks online status and does setup()"""

        # Is target online?
        if self.isOnline():
            # Setup our data.
            self.setup()
            return True

        # Not online.
        return False

    ###########################################################################
    #
    def isOnline(self):

        """Returns True if the target is online, False otherwise"""

        if self.online:
            return True

        self.sModel = getModel(self.sFullUrl , self.timeout)
        if self.sModel == 'Error':
            self.online = False
            self.ePlatform = Platform.BASC_NONE
            self.retryInterval = int(time.time())+self.retryInterval
        else:
            self.online = True
            self.ePlatform = getPlatform(self.sModel)
            self.retryInterval = 0
        return self.online

    ###########################################################################
    #
    def setup(self):

        """Sets device object quantities per the model"""

        # Get target platform.
        self.ePlatform = getPlatform(self.sModel)
        # Assign the number of IO.
        if self.ePlatform == Platform.BASC_20:
            self.uiQty = 8
            self.biQty = 4
            self.aoQty = 4
            self.boQty = 4
            self.vtQty = 24
        elif self.ePlatform == Platform.BASC_PI:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24
        elif self.ePlatform == Platform.BASC_AO:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 2
            self.boQty = 4
        elif self.ePlatform == Platform.BASC_PO:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24
        elif self.ePlatform == Platform.BASC_ED:
            self.uiQty = 6
            self.biQty = 0
            self.aoQty = 0
            self.boQty = 6
            self.vtQty = 24 
        elif self.ePlatform == Platform.BASC_EO:
             self.uiQty = 6
             self.biQty = 0
             self.aoQty = 2
             self.boQty = 4       
            # Assign 0-based index values.
        self.uiBase = 0
        self.biBase = self.uiQty
        self.aoBase = self.biBase + self.biQty
        self.boBase = self.aoBase + self.aoQty
        self.vtBase = self.boBase + self.boQty

    ###########################################################################
    #
    def readObject(self , iIndex):

        """Returns response text if success, None if failure"""

        # If not online.
        if not self.online:
            # If retry period has expired.
            if int(time.time()) >= self.nextRetry:
                # If we're still offline.
                if not self.isOnline():
                    # New retry time has been set.
                    return None
            # Else need to wait.
            else:
                return None

        # Assume failure.
        retval = None
        # XML read request string.
        req = '<rdom fcn="get" doc="rtd" path="obj[' + str(iIndex) + ']/val_scl"/>'
        # Try to execute the POST.
        try:
            rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
            # If good HTML response.
            if (rsp.status_code == 200):
                root = et.fromstring(rsp.text)
                if root.attrib['rsp'] == 'ack':
                    retval = root.text
            else:
                # Something there, but not BAScontrol.
                self.online = False
        except OSError:
            # Timeout.
            self.online = False
        # If we were taken offline.
        if not self.online:
            # Set for retry.
            self.nextRetry = int(time.time())+self.retryInterval
            # Raise an exception.
            ###raise IOError
            # NOT REACHED
        # And return.
        return retval 

    ###########################################################################
    #
    # This does two POST operations:
    #   1. Checks ws_control to see if it's under wire sheet control.
    #   2. Executes the write if not under wire sheet control.
    #
    def writeObject(self , iIndex , value):

        """Writes the present value of an object using 0-based index"""

        # If not online.
        if not self.online:
            # If retry period has expired.
            if int(time.time()) >= self.nextRetry:
                # If we're still offline.
                if not self.isOnline():
                    # New retry time has been set.
                    return -1
            # Else need to wait.
            else:
                return -1

        # Assume failure.
        retval = -1
        # XML wire sheet control request string.
        req = '<rdom fcn="get" doc="rtd" path="obj[' + str(iIndex) + ']/ws_control"/>'
        # Try to execute first POST request.
        try:
            # Execute POST and get response.
            rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
            # If good HTML response.
            if (rsp.status_code == 200):
                # Parse XML.
                root = et.fromstring(rsp.text)
                # If request was ACK'd.
                ###if root.attrib['rsp'] == 'ack': ### Allow Wire Sheet Writes
                    # If point is under wire sheet control.
                ###    if root.text == '1':
                        # We can't write.
                ###        return -1
                # Else we were NAK'd; SERIOUS PROBLEM
                ###else:
                ###    return -1
                #
                # If we get here, we can proceed with the write operation.
                #
                # XML write request string.
                req = '<rdom fcn="set" doc="rtd" path="obj[' + str(iIndex) + ']/val_scl">' + str(value) + '</rdom>'
                # Try to execute POST request.
                try:
                    # Execute POST and get response.
                    rsp = requests.post(self.sFullUrl, data=req , timeout=self.timeout)
                    # If good HTML response.
                    if (rsp.status_code == 200):
                        # Parse XML.
                        root = et.fromstring(rsp.text)
                        #######################################################
                        # This is the only place we can return success.
                        #######################################################
                        if root.attrib['rsp'] == 'ack':
                            retval = 0
                    else:
                        # Something there, but not BAScontrol.
                        self.online = False
                except OSError:
                    # Timeout.
                    self.online = False
            else:
                # Something there, but not BAScontrol.
                self.online = False
        except OSError:
            # Timeout.
            self.online = False
        # If we were taken offline.
        if not self.online:
            # Set for retry.
            self.nextRetry = int(time.time())+self.retryInterval
            ### Raise an exception.
            ###raise IOError
            # NOT REACHED
        # And return.
        return retval

    ###########################################################################
    #
    def universalInput(self , iIndex):

        """Reads a universal input using 1-based index (UI1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.uiQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1)
        # Execute read.
        value = self.readObject(index)
        # And return.
        return None if value == None else float(value)
        

    ###########################################################################
    #
    def binaryInput(self , iIndex):

        """Reads a binary input using 1-based index (BI1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.biQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.biBase)
        # Execute read.
        value = self.readObject(index)
        # And return.
        return None if value == None else int(value)

    ###########################################################################
    #
    def analogOutput(self , iIndex , fValue = None):

        """Reads or writes an analog output using 1-based index (AO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.aoQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.aoBase)
        # If reading.
        if fValue == None:
            # Execute read.
            value = self.readObject(index)
            # And return.
            return None if value == None else float(value)
        # Else writing.
        return self.writeObject(index , fValue)

    ###########################################################################
    #
    def binaryOutput(self, iIndex, iValue=None):

        """Reads or writes an binary output using 1-based index (BO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.boQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.boBase)
        # If reading.
        if iValue == None:
            # Execute read.
            value = self.readObject(index)
            # Return if failure.
            if value == None:
                return None
            # Return with integer value.
            return 0 if int(value) == 0 else 1
        # Else writing.
        return self.writeObject(index, iValue)

    ###########################################################################
    #
    def virtualValue(self , iIndex , bIsBinary , aValue = None):

        """Reads or writes a virtual point using 1-based index (BO1..)"""

        # Sanity check.
        if (iIndex <= 0) or (iIndex > self.vtQty) or (self.ePlatform == Platform.BASC_NONE):
            return None
        # Get 0-based index.
        index = int(iIndex - 1 + self.vtBase)
        # If reading.
        if aValue == None:
            # Execute read.
            value = self.readObject(index)
            # Return if failure.
            if value == None:
                return None
            # Return with value.
            return float(value) if bIsBinary else float(value)
        # Else writing.
        return self.writeObject(index , aValue)

# contemporary controls / Universal devices, sjb gtb 7/2019
# Edited and re-vamped slightly on 8_19_2020 to remove IO error stopping Nodeserver